### PR TITLE
RATEST-275:NEW E2E Test: Manage OCL Subscription Workflow

### PR DESCRIPTION
**Pull request by jwnasambu**
Description of what I did:

I created the following files:

-  open-Concept-Lab.yml file
- made changes on package.json file
- modified ConfigureMetadataPage
- Created OpenConceptLabPage
- Created OpenConceptLabSuccessPage
- Created SubscriptionPage
- Modified RunTest
- Created OpenConceptLabSteps
- Created openConceptLab.feature

Issue I worked on:

https://issues.openmrs.org/browse/RATEST-275

### CODE REVIEW

**Comment by jwnasambu** [comment]
This the reference ticket https://github.com/openmrs/openmrs-contrib-qaframework/pull/280

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+     #user story:Edit subscription
+     When a user clicks edit subscription button
+    Then the system loads Subscription page
+      And a user edits the URL of a released dictionary
+      And a user edits the Token url
+      And a user clicks on the Save Changes button
+      And the system loads Open Concept Lab page
+      And a user clicks import from Subscription server button
+     Then the API should be displayed on the previous imports
+     #user story: Unsubscribe
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:30]:
```js
 #user story: Unsubscribe
     When a user clicks edit subscription button
     Then the system loads Subscription page
      And a user clicks unsubscribe button
     Then the system loads Subscription page 
```
Does it require one to click in edit subscription button in order to unsubscribe?

Comment (jwnasambu) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:30]: @kdaud yes.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+     Then the system loads Subscription page
+      And a user enters the URL of a new released dictionary
+      And a user enters the Token url
+      And a user clicks on the Save Changes button
+     Then the system loads Open Concept Lab Success page
+      And a user clicks import from Subscription server button
+      Then the API should be displayed on the previous imports
+     #user story:Edit subscription
+     When a user clicks edit subscription button
+    Then the system loads Subscription page
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:23]: Put all the feature files on the same line.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import org.openmrs.contrib.qaframework.page.OpenConceptLabPage;
+import org.openmrs.contrib.qaframework.page.OpenConceptLabSuccessPage;
+import org.openmrs.contrib.qaframework.page.SubscriptionPage;
+
+import io.cucumber.java.After;
+import io.cucumber.java.Before;
+import io.cucumber.java.en.And;
+import io.cucumber.java.en.Given;
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java:26]: checkout the import order in the existing files and align!

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature (hunk 1 above)
Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:30]: What is the user story accomplishing?

Comment (jwnasambu) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:30]: @kdaud Kindly you can try it out on the reference application to get a clear picture but without it the system can't run normally much less when the information is in the stored in the database.

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:30]: What is the user story all about?

Comment (jwnasambu) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:30]: @kdaud Kindly I don't seem to understand what you are asking neither where you are heading. Do you mind being clear and to the point on what you are trying to put across?

Comment (kdaud) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:30]: The user story 
```js
#user story: Unsubscribe
```
what does it accomplish? In other words what can a user do with the feature?

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/OpenConceptLabSuccessPage.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.util.List;
+
+import org.openmrs.contrib.qaframework.helper.Page;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+
+public class OpenConceptLabSuccessPage extends Page {
+
+    private static final String PAGE_URL = "/owa/openconceptlab/index.html#/";
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/OpenConceptLabSuccessPage.java:22]:
```suggestion
    private static final String PAGE_URL = "/owa/openconceptlab/index.html";
```

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/OpenConceptLabPage.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ */
+package org.openmrs.contrib.qaframework.page;
+
+import org.openmrs.contrib.qaframework.helper.Page;
+import org.openqa.selenium.By;
+
+public class OpenConceptLabPage extends Page {
+	
+    private static final String PAGE_URL = "/owa/openconceptlab/index.html#/";
+    private static final By SETUP_SUBSCRIPTION_BUTTON = By.linkText("Setup Subscription");
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/OpenConceptLabPage.java:18]: You can add space to separate attributes and constructor method.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    private OpenConceptLabSuccessPage openConceptLabSuccessPage;
+    private SubscriptionPage subscriptionPage;
+    
+    @Before(RunTest.HOOK.SELENIUM_OPEN_CONCEPT_LAB)
+    public void visitHomePage() {
+        initiateWithLogin();    
+    }
+    
+    @After(RunTest.HOOK.SELENIUM_OPEN_CONCEPT_LAB)
+    public void destroy(){
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java:43]:
```suggestion
    public void destroy() {
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        assertTrue(textExists("Show/hide advanced..."));
+    }
+
+    @And("a user enters the URL of a new released dictionary")
+    public void enterDictionaryUrl() {
+        subscriptionPage.enterSubscriptionURL(releasedDictionaryUrl);
+    }   
+
+    @And("a user enters the Token url")
+    public void enterTokenUrl(){
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java:83]:
```suggestion
    public void enterTokenUrl() {
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+         subscriptionPage = (SubscriptionPage) openConceptLabSuccessPage.clickEditSubscriptionButton();  
+     } 
+    
+    @And("a user edits the URL of a released dictionary")
+    public void editDictionaryUrl() {
+        subscriptionPage.enterSubscriptionURL("https://api.qa-refapp.openmrs.org/openmrs/dictionary/concept.htm?conceptId=162172");
+    }   
+
+    @And("a user edits the Token url")
+    public void editTokenUrl(){
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java:119]:
```suggestion
    public void editTokenUrl()  {
```
Check all these similar changes and correct them appropriately.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
+
+import org.openmrs.contrib.qaframework.RunTest;
+import org.openmrs.contrib.qaframework.page.ConfigureMetadataPage;
+import org.openmrs.contrib.qaframework.page.OpenConceptLabPage;
+import org.openmrs.contrib.qaframework.page.OpenConceptLabSuccessPage;
+import org.openmrs.contrib.qaframework.page.SubscriptionPage;
+ 
+public class OpenConceptLabSteps extends Steps  {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java:28]:
```suggestion
public class OpenConceptLabSteps extends Steps {
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+	
+    private OpenConceptLabPage openconceptlabpage;
+    private ConfigureMetadataPage configureMetadataPage;
+    private static final String releasedDictionaryUrl = "https://api.staging.openconceptlab.org/users/openmrs/collections/TD-ecly/65725685/";
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java:32]: Align with the naming of the final variables!

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    private static final String tokenUrl = "bd022mad6d3df24f5c42ewewa94b53a23edf6eee7r";
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java:33]: How is the `tokenUrl` generated?

**Comment by jwnasambu** [comment]
@kdaud The token URL is personal and It's the one you have in your OCL account.

**Comment by kdaud** [comment]
> The token URL is personal and It's the one you have in your OCL account

Could you direct me where this token is stored?

**Comment by jwnasambu** [comment]
@kdaud kindly get the clear information on this link  https://talk.openmrs.org/t/automated-tests-for-the-ocl-module/34657/16 on how to go about it.

**Comment by kdaud** [comment]
I have gone through the thread and did not get directive where the token url is stored. 

Though the point is:'

If the tokenUrl is generated at run time then it is will not be changing on every run so the test will certainly break.
If it's a static URL then all is well and if so where can I get it?

**Comment by jwnasambu** [comment]
@kdaud Kindly I shared with you the link so that you can have a clear insight on what I am working on. Concerning the Token URL, after the explanation from Michael, I decided to research on the Token URL on TALK and I landed on this [post](https://talk.openmrs.org/t/ocl-subscription-using-openmrs-module-openconceptlab/14537/4) which lead me to the [staging server](https://app.staging.openconceptlab.org/) and I followed these step to get the Token-url.


Uploading screencast 2022-03-03 12 AM-04-50.mp4…

cc @ibacher,

**Comment by jwnasambu** [comment]
https://user-images.githubusercontent.com/33891016/156454467-75df3539-d932-4e74-b6dc-d564ef498794.mp4

**Comment by kdaud** [comment]
@jwnasambu is the tokenUrl a static one? In other wards does the token remain the same when you login into the server several times?

**Comment by ibacher** [comment]
> is the tokenUrl a static one?

Yes, the OCL tokens are static. One of the reasons I strongly prefer using GitHub Secrets for them.

**Comment by kdaud** [comment]
@ibacher have logged into ocl server and I get a different token than the one suggested in this commit. Logging into the server again I see a different token generated and this seems to suggest that the tokenUrl is generated at run time and if so this will cause the test to fail.

Let me know what is your thought?

**Comment by ibacher** [comment]
@kdaud IIRC OCL tokens get regenerated every time the password is reset, but otherwise, they are stable for a user on an instance.

**Comment by kdaud** [comment]
> @kdaud IIRC OCL tokens get regenerated every time the password is reset, but otherwise, they are stable for a user on an instance.

This makes sense!

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+Background:
+    Given a user clicks on configure Metadata link from home page
+    When the system loads the configure metadata page
+    And a user clicks Manage OCL link from configure metadata page
+    Then the system loads Open Concept Lab page
+
+  @selenium
+  @openConceptLab
+  Scenario: OCL MANAGEMENT
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:11]:
```suggestion
  Scenario: OCL Management```

Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/16-openConceptLab/openConceptLab.feature:11]:
```suggestion
  Scenario: OCL Management
```

**Comment by kdaud** [comment]
Thanks @jwnasambu for the hard work, have left you with one more review for you!

**Comment by kdaud** [comment]
@jwnasambu do you  still need this PR to be as a draft?

**Comment by jwnasambu** [comment]
@kdaud am so sorry it had not clicked my mind .

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @And("a user clicks import from Subscription server button")
+    public void LoadsOpenConceptLabPage() {
+        openConceptLabSuccessPage.clickOnImportFromSubscriptionButton();
+        openConceptLabSuccessPage.waitForPageToLoad();
+    }
+
+    @Then("the API should be displayed on the previous imports")
+    public void displayAmongPreviousImports() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java:104]: is the method name reflecting what is being performed?

**Comment by jwnasambu** [comment]
@kdaud when you click on `import from subscription server` on this page https://qa-refapp.openmrs.org/openmrs/owa/openconceptlab/index.html#/ you should get this output
Screenshot 2022-03-08 at 9 30 04 PM Of which the API is the status.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java (hunk 13 above)
Comment (kdaud) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/OpenConceptLabSteps.java:104]: @jwnasambu take a close look at the method name.

**Comment by jwnasambu** [comment]
@kdaud am sorry I haven't the PR because am having issues with my testing environment on both chrome and firefox browser. Yesterday I had a call with @sherrif10  for over an hour but still we didn't succeed. This is the issue am facing https://talk.openmrs.org/t/found-argument-webdriver-port-which-wasnt-expected-or-isnt-valid-in-this-context/35936/12

**Comment by jwnasambu** [comment]
@kdaud, @ibacher  and @sherrif10 kindly this is the output I have as the subscription module workflow is concerned

Uploading screencast 2022-03-10 2 AM-16-14.mp4….

File: .github/workflows/refapp-2x-open-Concept-Lab.yml [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+        run: |
+          docker-compose -f docker/docker-compose-refqa.yml up -d
+      - name: wait for openmrs instance to start
+        run: while [[ "$(curl -s -o /dev/null -w ''%{http_code}'' http://localhost:8080/openmrs/login.htm)" != "200" ]]; do sleep 1; done
+      - name: Run qaframework on selenium
+        run: |
+          echo "cucumber.publish.enabled=true" > qaframework-bdd-tests/src/test/resources/cucumber.properties
+          mvn clean install -DskipTests=true
+          cp -f qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test.properties
+          cd qaframework-bdd-tests
```
Comment (kdaud) [inline .github/workflows/refapp-2x-open-Concept-Lab.yml:38]: We can rather set `qaframework-bdd-tests` as a default working directory for all the commands to run within it, cross check the existing .yml file to see how it's done?

**Comment by jwnasambu** [comment]
@kdaud kindly shade some  more light on your request above please!

**Comment by kdaud** [comment]
qaframework-bdd-tests directory should be set as a  default  one for all the commands, take a look at any of the existing RefApp 2.x .yml files.

**Comment by jwnasambu** [comment]
@kdaud kindly am not seeing the changes you are requesting for. I have cross checked with other RefApp 2.x.yml files and the content seems to be the same. It has been an hour now without a success and I have a lot on my plate, do you mind pointing out directly what am missing please?

**Comment by kdaud** [comment]
Take a look at https://github.com/openmrs/openmrs-contrib-qaframework/blob/6322cc966f3fd0ce8692ba58d315b356d5039e59/.github/workflows/refapp-2x-vitals-and-triaging.yml#L1-L42 and compare with your .yml file and align.

**Comment by jwnasambu** [comment]
@kdaud since I couldn't see any difference I just edited the file you shared.

**Comment by kdaud** [comment]
@jwnasambu have you taken a look at the build failures?

File: .github/workflows/refapp-2x-open-Concept-Lab.yml [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
```
Comment (kdaud) [inline .github/workflows/refapp-2x-open-Concept-Lab.yml:38]: We can rather set `qaframework-bdd-tests` as a default working directory for all the commands to run within it, cross check the existing .yml file to see how it's done.

**Comment by jwnasambu** [comment]
@kdaud kindly when you get time let us have a sync please!

**Comment by jwnasambu** [comment]
@kdaud thanks so so much for the clear explanation I can now gladly fix the proposed changes because I know what is required.

**Pull request closed without merging**